presto-jdbc/src/main/java/com/facebook/presto/jdbc/ReferenceCountingLoadingCache.java

```diff
@@ -0,0 +1,214 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.jdbc;
+
+import com.google.common.base.Ticker;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import com.google.common.cache.RemovalListener;
+import com.google.common.cache.RemovalNotification;
+
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Consumer;
+
+import static com.google.common.base.Preconditions.checkState;
+import static io.airlift.concurrent.Threads.daemonThreadsNamed;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+class ReferenceCountingLoadingCache<K, V>
+{
+    private static class Holder<V>
+    {
+        private final V value;
+        private int refcount;
+        private int cleanerCount;
+        private ScheduledFuture currentCleanup;
+
+        private Holder(V value)
+        {
+            this.value = requireNonNull(value, "value is null");
+        }
+
+        private V get()
+        {
+            return value;
+        }
+
+        private int getRefcount()
+        {
+            return refcount;
+        }
+
+        private int getCleanerCount()
+        {
+            return cleanerCount;
+        }
+
+        private int reference()
```

Comment:
I was shooting for something that didn't collide with acquire/release in the cache itself (the code is confusing enough without having the same names meaning different things). Got a suggestion other than release?